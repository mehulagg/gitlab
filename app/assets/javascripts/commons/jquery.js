import 'jquery';

// common jQuery plugins
import 'jquery.caret'; // must be imported before at.js
import '@gitlab/at.js';
import 'vendor/jquery.scrollTo';
