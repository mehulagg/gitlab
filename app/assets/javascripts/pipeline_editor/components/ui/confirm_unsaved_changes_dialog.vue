<script>
export default {
  props: {
    hasUnsavedChanges: {
      type: Boolean,
      required: true,
    },
  },
  created() {
    window.addEventListener('beforeunload', this.confirmChanges);
  },
  destroyed() {
    window.removeEventListener('beforeunload', this.confirmChanges);
  },
  methods: {
    confirmChanges(e) {
      if (this.hasUnsavedChanges) {
        e.preventDefault();
        e.returnValue = '';
      }
    },
  },
  render: () => null,
};
</script>
