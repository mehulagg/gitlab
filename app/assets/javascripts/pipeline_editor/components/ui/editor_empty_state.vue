<script>
import { GlSprintf } from '@gitlab/ui';
import { s__ } from '~/locale';

export default {
  components: {
    GlSprintf,
  },
  i18n: {
    title: s__('PipelineEditor|Optimize your workflow with CI/CD Pipelines'),
    body: s__(
      'PipelineEditor|Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center">
    <img :src="emptyStateIllustrationPath" data-testid="empty-state-illustration" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p>
      <gl-sprintf :message="$options.i18n.body" class="gl-display-inline">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
  </div>
</template>
