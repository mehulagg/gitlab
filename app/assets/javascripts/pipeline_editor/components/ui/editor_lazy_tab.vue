<script>
import { GlTab } from '@gitlab/ui';

/**
 * Wrapper of <gl-tab> to lazily render this tab's content
 * when shown **without dismounting it after**.
 *
 * Once the tab is selected it is permanently set as "not-lazy"
 * so it's contents are not dismounted.
 *
 * Usage:
 *
 * Same as <gl-tab>, for example:
 *
 * <gl-tabs>
 *   <editor-lazy-tab title="Tab 1"> ... </editor-lazy-tab>
 *   <editor-lazy-tab title="Tab 2"> ... </editor-lazy-tab>
 * </gl-tabs>
 */
export default {
  components: {
    GlTab,
  },
  props: {
    lazy: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  data() {
    return {
      isLazy: this.lazy,
    };
  },
  methods: {
    onClick() {
      this.isLazy = false;
    },
  },
};
</script>
<template>
  <gl-tab :lazy="isLazy" v-bind="$attrs" v-on="$listeners" @click="onClick">
    <slot v-for="slot in Object.keys($slots)" :slot="slot" :name="slot"></slot>
  </gl-tab>
</template>
