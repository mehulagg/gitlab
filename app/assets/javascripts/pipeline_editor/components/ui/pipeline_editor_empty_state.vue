<script>
import { GlButton, GlSprintf } from '@gitlab/ui';
import { __ } from '~/locale';
import glFeatureFlagsMixin from '~/vue_shared/mixins/gl_feature_flags_mixin';

export default {
  components: {
    GlButton,
    GlSprintf,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Create new CI/CD pipeline'),
  },
  mixins: [glFeatureFlagsMixin()],
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
  computed: {
    showCTAButton() {
      return this.glFeatures.pipelineEditorEmptyStateAction;
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div class="gl-display-flex gl-flex-direction-column gl-align-items-center gl-mt-11">
    <img :src="emptyStateIllustrationPath" />
    <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
    <p class="gl-mt-3">
      <gl-sprintf :message="$options.i18n.body">
        <template #code="{ content }">
          <code>{{ content }}</code>
        </template>
      </gl-sprintf>
    </p>
    <gl-button
      v-if="showCTAButton"
      variant="confirm"
      class="gl-mt-3"
      @click="createEmptyConfigFile"
    >
      {{ $options.i18n.btnText }}
    </gl-button>
  </div>
</template>
