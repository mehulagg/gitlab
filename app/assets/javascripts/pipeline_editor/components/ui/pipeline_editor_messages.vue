<script>
import { GlAlert } from '@gitlab/ui';
import { getParameterValues, removeParams } from '~/lib/utils/url_utility';
import { __, s__ } from '~/locale';
import {
  COMMIT_FAILURE,
  COMMIT_SUCCESS,
  DEFAULT_FAILURE,
  LOAD_FAILURE_UNKNOWN,
} from '../../constants';
import CodeSnippetAlert from '../code_snippet_alert/code_snippet_alert.vue';
import {
  CODE_SNIPPET_SOURCE_URL_PARAM,
  CODE_SNIPPET_SOURCES,
} from '../code_snippet_alert/constants';

export default {
  components: {
    GlAlert,
    CodeSnippetAlert,
  },
  errorTexts: {
    [COMMIT_FAILURE]: s__('Pipelines|The GitLab CI configuration could not be updated.'),
    [DEFAULT_FAILURE]: __('Something went wrong on our end.'),
    [LOAD_FAILURE_UNKNOWN]: s__('Pipelines|The CI configuration was not loaded, please try again.'),
  },
  successTexts: {
    [COMMIT_SUCCESS]: __('Your changes have been successfully committed.'),
  },
  props: {
    failureType: {
      type: String,
      required: false,
      default: null,
    },
    failureReasons: {
      type: Array,
      required: false,
      default: () => [],
    },
    successType: {
      type: String,
      required: false,
      default: null,
    },
  },
  data() {
    return {
      codeSnippetCopiedFrom: '',
      showAlert: false,
    };
  },
  computed: {
    alerts() {
      const messages = [];

      if (this.failureType) {
        messages.push(this.failure);
      }

      if (this.success) {
        messages.push(this.success);
      }
      return messages;
    },
    failure() {
      switch (this.failureType) {
        case LOAD_FAILURE_UNKNOWN:
          return {
            text: this.$options.errorTexts[LOAD_FAILURE_UNKNOWN],
            variant: 'danger',
          };
        case COMMIT_FAILURE:
          return {
            text: this.$options.errorTexts[COMMIT_FAILURE],
            variant: 'danger',
          };
        default:
          return {
            text: this.$options.errorTexts[DEFAULT_FAILURE],
            variant: 'danger',
          };
      }
    },
  },
  success() {
    switch (this.successType) {
      case COMMIT_SUCCESS:
        return {
          text: this.$options.successTexts[COMMIT_SUCCESS],
          variant: 'info',
        };
      default:
        return null;
    }
  },

  created() {
    this.parseCodeSnippetSourceParam();
  },
  methods: {
    dismissAlert() {
      this.showAlert = false;
    },
    dismissCodeSnippetAlert() {
      this.codeSnippetCopiedFrom = '';
    },
    parseCodeSnippetSourceParam() {
      const [codeSnippetCopiedFrom] = getParameterValues(CODE_SNIPPET_SOURCE_URL_PARAM);
      if (codeSnippetCopiedFrom && CODE_SNIPPET_SOURCES.includes(codeSnippetCopiedFrom)) {
        this.codeSnippetCopiedFrom = codeSnippetCopiedFrom;
        window.history.replaceState(
          {},
          document.title,
          removeParams([CODE_SNIPPET_SOURCE_URL_PARAM]),
        );
      }
    },
  },
};
</script>
<template>
  <div>
    <code-snippet-alert
      v-if="codeSnippetCopiedFrom"
      :source="codeSnippetCopiedFrom"
      class="gl-mb-5"
      @dismiss="dismissCodeSnippetAlert"
    />
    <gl-alert
      v-for="alert in alerts"
      :key="alert.text"
      :variant="alert.variant"
      class="gl-mb-5"
      @dismiss="dismissAlert"
    >
      {{ alert.text }}
      <ul v-if="failureReasons.length" class="gl-mb-0">
        <li v-for="reason in failureReasons" :key="reason">{{ reason }}</li>
      </ul>
    </gl-alert>
  </div>
</template>
