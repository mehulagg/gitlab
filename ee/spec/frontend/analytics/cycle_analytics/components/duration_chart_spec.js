import { shallowMount, mount } from '@vue/test-utils';
import { GlLoadingIcon } from '@gitlab/ui';
import $ from 'jquery';
import 'bootstrap';
import '~/gl_dropdown';
import Scatterplot from 'ee/analytics/shared/components/scatterplot.vue';
import DurationChart from 'ee/analytics/cycle_analytics/components/duration_chart.vue';
import StageDropdownFilter from 'ee/analytics/cycle_analytics/components/stage_dropdown_filter.vue';
import {
  allowedStages as stages,
  durationChartPlottableData as scatterData,
  durationChartPlottableMedianData as medianLineData,
} from '../mock_data';

function createComponent({ mountFn = shallowMount, props = {}, stubs = {} } = {}) {
  return mountFn(DurationChart, {
    propsData: {
      isLoading: false,
      stages,
      scatterData,
      medianLineData,
      ...props,
    },

    stubs: {
      GlLoadingIcon: true,
      Scatterplot: true,
      StageDropdownFilter: true,
      ...stubs,
    },
  });
}

describe('DurationChart', () => {
  let wrapper;

  const findNoDataContainer = _wrapper => _wrapper.find({ ref: 'duration-chart-no-data' });
  const findScatterPlot = _wrapper => _wrapper.find(Scatterplot);
  const findStageDropdown = _wrapper => _wrapper.find(StageDropdownFilter);
  const findLoader = _wrapper => _wrapper.find(GlLoadingIcon);

  const openStageDropdown = _wrapper => {
    $(findStageDropdown(_wrapper).element).trigger('shown.bs.dropdown');
    return _wrapper.vm.$nextTick();
  };

  const selectStage = (_wrapper, index = 0) => {
    findStageDropdown(_wrapper)
      .findAll('a')
      .at(index)
      .trigger('click');
    return _wrapper.vm.$nextTick();
  };

  beforeEach(() => {
    wrapper = createComponent({});
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders the duration chart', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders the scatter plot', () => {
    expect(findScatterPlot(wrapper).exists()).toBe(true);
  });

  it('renders the stage dropdown', () => {
    expect(findStageDropdown(wrapper).exists()).toBe(true);
  });

  describe('when a stage is selected', () => {
    const selectedIndex = 1;
    const selectedStages = stages.filter((_, index) => index !== selectedIndex);

    beforeEach(() => {
      wrapper = createComponent({ mountFn: mount, stubs: { StageDropdownFilter: false } });
      return openStageDropdown(wrapper).then(() => selectStage(wrapper, selectedIndex));
    });

    it('emits the stageSelected event', () => {
      expect(wrapper.emitted().stageSelected).toBeTruthy();
    });

    it('toggles the selected stage', () => {
      expect(wrapper.emitted('stageSelected')[0]).toEqual([selectedStages]);

      return selectStage(wrapper, selectedIndex).then(() => {
        const [updatedStages] = wrapper.emitted('stageSelected')[1];
        stages.forEach(stage => {
          expect(updatedStages).toContain(stage);
        });
      });
    });
  });

  describe('with no chart data', () => {
    beforeEach(() => {
      wrapper = createComponent({ props: { scatterData: [], medianLineData: [] } });
    });

    it('renders the no data available message', () => {
      expect(findNoDataContainer(wrapper).text()).toEqual(
        'There is no data available. Please change your selection.',
      );
    });
  });

  describe('while loading', () => {
    beforeEach(() => {
      wrapper = createComponent({ props: { isLoading: true } });
    });
    it('renders loading icon', () => {
      expect(findLoader(wrapper).exists()).toBe(true);
    });
  });
});
