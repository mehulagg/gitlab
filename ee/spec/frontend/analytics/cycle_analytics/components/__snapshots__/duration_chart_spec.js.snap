// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationChart renders the duration chart 1`] = `
"<div>
  <div class=\\"gl-display-flex\\">
    <h4 class=\\"gl-mt-0\\">Days to completion</h4>
    <stagedropdownfilter-stub stages=\\"[object Object],[object Object],[object Object]\\" label=\\"stage dropdown\\" class=\\"gl-ml-auto\\"></stagedropdownfilter-stub>
  </div>
  <scatterplot-stub xaxistitle=\\"Date\\" yaxistitle=\\"Total days to completion\\" scatterdata=\\"2019-01-01,29,2019-01-01,2019-01-02,100,2019-01-02\\" medianlinedata=\\"2018-12-31,29,2019-01-01,100\\" tooltipdateformat=\\"mmm d, yyyy\\"></scatterplot-stub>
</div>"
`;
